app/models/message.rb

```diff
@@ -13,59 +14,45 @@ class Message < ActiveRecord::Base
   has_many :attachments, dependent: :destroy
   has_many :log_items, -> { where(loggable_type: 'Message') }, foreign_key: 'loggable_id'
   belongs_to :site
+  has_and_belongs_to_many :members, class_name: 'Person', join_table: :members_messages, association_foreign_key: :member_id
 
   scope_by_site_id
-
-  scope :same_as, -> m { where('id != ?', m.id || 0).where(person_id: m.person_id, subject: m.subject, body: m.body, to_person_id: m.to_person_id, group_id: m.group_id).where('created_at >= ?', 1.day.ago) }
+  scope(:same_as, lambda do |m|
+    where('id != ?', m.id || 0)
+      .where(person_id:    m.person_id,
+             subject:      m.subject,
+             body:         m.body,
+             to_person_id: m.to_person_id,
+             group_id:     m.group_id)
+      .where('created_at >= ?', 1.day.ago)
+  end)
+  scope(:for_user, lambda do |user|
+    joins(:members).where('members_messages.member_id = ? OR messages.person_id = ?', user.id, user.id).uniq 
+  end)
+  scope :for_whole_group, -> { includes(:members).where(members_messages: { :message_id => nil }) }
 
   validates_presence_of :person_id
-  validates_presence_of :subject
   validates_length_of :subject, minimum: 2
-
-  validates_each :to_person_id, allow_nil: true do |record, attribute, value|
-    if attribute.to_s == 'to_person_id' and value and record.to and record.to.email.nil?
-      record.errors.add attribute, :invalid
-    end
-  end
-
-  validates_each :body do |record, attribute, value|
-    if attribute.to_s == 'body' and value.to_s.blank? and record.html_body.to_s.blank?
-      record.errors.add attribute, :blank
-    end
-  end
-
-  def top
-    top = self
-    while top.parent
-      top = top.parent
-    end
-    return top
+  validate do |message|
+    errors.add(:to, :invalid) if message.to && message.to.email.nil?
+    errors.add(:body, :blank) unless body || html_body
   end
 
-  before_save :remove_unsubscribe_link
-
-  def remove_unsubscribe_link
-    if body
-      body.gsub!(/http:\/\/.*?person_id=\d+&code=\d+/i, '--removed--')
-    end
-    if html_body
-      html_body.gsub!(/http:\/\/.*?person_id=\d+&code=\d+/i, '--removed--')
-    end
+  def top(top_message = self)
+    top_message.parent ? top(top_message.parent) : top_message
   end
 
-  before_save :remove_message_id_in_body
-
-  def remove_message_id_in_body
-    if body
-      body.gsub! MESSAGE_ID_RE_IN_BODY, ''
-    end
-    if html_body
-      html_body.gsub! MESSAGE_ID_RE_IN_BODY, ''
-    end
+  before_save do
+    self.body &&= body.gsub(%r{http://.*?person_id=\d+&code=\d+}i, '--removed--').gsub(MESSAGE_ID_RE_IN_BODY, '')
+    self.html_body &&= html_body.gsub(%r{http://.*?person_id=\d+&code=\d+}i, '--removed--')
+                                .gsub(MESSAGE_ID_RE_IN_BODY, '')
+    self.code ||= rand(999_999) + 1
+    self.member_ids = parent.member_ids << parent.person_id if group_id? && parent_id? && parent.member_ids.any?
```

Comment:
I appreciate you consolidating code, but I think the readability went way down because these are no longer separate named methods. `before_save :some_meaningful_name` is better.